Make cudaHostRegister actually useful on cudart.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45159 Make cudaHostRegister actually useful on cudart.**

By default, pybind11 binds void* to be capsules.  After a lot of
Googling, I have concluded that this is not actually useful:
you can't actually create a capsule from Python land, and our
data_ptr() function returns an int, which means that the
function is effectively unusable.  It didn't help that we had no
tests exercising it.

I've replaced the void* with uintptr_t, so that we now accept int
(and you can pass data_ptr() in directly).  I'm not sure if we
should make these functions accept ctypes types; unfortunately,
pybind11 doesn't seem to have any easy way to do this.

Fixes #43006

Also added cudaHostUnregister which was requested.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D23849731](https://our.internmc.facebook.com/intern/diff/D23849731)

ezyang committed:
Make cudaHostRegister actually useful on cudart.

By default, pybind11 binds void* to be capsules.  After a lot of
Googling, I have concluded that this is not actually useful:
you can't actually create a capsule from Python land, and our
data_ptr() function returns an int, which means that the
function is effectively unusable.  It didn't help that we had no
tests exercising it.

I've replaced the void* with uintptr_t, so that we now accept int
(and you can pass data_ptr() in directly).  I'm not sure if we
should make these functions accept ctypes types; unfortunately,
pybind11 doesn't seem to have any easy way to do this.

Fixes #43006

Also added cudaHostUnregister which was requested.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

lw commented on test/test_cuda.py:
```diff
@@ -279,6 +279,14 @@ def assert_change(comp=1, empty_cache=False, reset_peak=False):
+        cudart = torch.cuda.cudart()
+        r = cudart.cudaHostRegister(t.data_ptr(), t.numel() * t.element_size(), 0)
+        self.assertEquals(r, 0)
+        r = cudart.cudaHostUnregister(t.data_ptr())
```
Add `self.assertTrue(t.is_pinned())`?

ezyang replied:
Done!

lw reviewed (approved):
Awesome, thanks!

ezyang committed:
Update on "Make cudaHostRegister actually useful on cudart."


By default, pybind11 binds void* to be capsules.  After a lot of
Googling, I have concluded that this is not actually useful:
you can't actually create a capsule from Python land, and our
data_ptr() function returns an int, which means that the
function is effectively unusable.  It didn't help that we had no
tests exercising it.

I've replaced the void* with uintptr_t, so that we now accept int
(and you can pass data_ptr() in directly).  I'm not sure if we
should make these functions accept ctypes types; unfortunately,
pybind11 doesn't seem to have any easy way to do this.

Fixes #43006

Also added cudaHostUnregister which was requested.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D23849731](https://our.internmc.facebook.com/intern/diff/D23849731)

[ghstack-poisoned]

lw reviewed (approved):
Perfect!

ezyang committed:
Update on "Make cudaHostRegister actually useful on cudart."


By default, pybind11 binds void* to be capsules.  After a lot of
Googling, I have concluded that this is not actually useful:
you can't actually create a capsule from Python land, and our
data_ptr() function returns an int, which means that the
function is effectively unusable.  It didn't help that we had no
tests exercising it.

I've replaced the void* with uintptr_t, so that we now accept int
(and you can pass data_ptr() in directly).  I'm not sure if we
should make these functions accept ctypes types; unfortunately,
pybind11 doesn't seem to have any easy way to do this.

Fixes #43006

Also added cudaHostUnregister which was requested.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D23849731](https://our.internmc.facebook.com/intern/diff/D23849731)

[ghstack-poisoned]
